File: ruby/problem1.rb
require 'benchmark/ips'
require 'minitest/autorun'

module Euler
  class ProblemOne
    # two pass iterative
    def solve n
      (1..n).select { |n| n % 3 < 1 || n % 5 < 1 }.inject(:+)
    end

    # single pass recursive
    def rec_solve n
      return 0 if n < 3
      sum = n % 3 < 1 || n % 5 < 1 ? n : 0
      sum + rec_solve(n - 1)
    end

    # single pass iterative
    def itr_solve n
      sum = 0
      while n > 2
        sum = sum + n if n % 3 < 1 || n % 5 < 1
        n = n - 1
      end
      sum
    end

    def benchmark
      Benchmark.ips do |x|
        x.report('base') { solve 999 }
        x.report('itr') { itr_solve 999 }
        x.report('rec') { rec_solve 999 }
        x.compare!
      end
    end
  end

  class ProblemOneTest < Minitest::Test
    def setup
      @problem = ProblemOne.new
    end

    def test_solution
      assert_equal @problem.solve(999), 233_168
    end

    def test_recursive_solution
      assert_equal @problem.rec_solve(999), 233_168
    end

    def test_iterative_solution
      assert_equal @problem.itr_solve(999), 233_168
    end
  end
end

puts Euler::ProblemOne.new.benchmark

File: elixir/problem1.exs
defmodule Euler.ProblemOne do
  def sum_of_threes_and_fives(n) when n < 3 do
    0
  end

  def sum_of_threes_and_fives(n) when rem(n,3) < 1 or rem(n,5) < 1 do
    n + sum_of_threes_and_fives(n - 1)
  end

  def sum_of_threes_and_fives(n) do
    sum_of_threes_and_fives(n - 1)
  end
end

ExUnit.start
defmodule ProblemOneTest do
  use ExUnit.Case, async: true

  test 'solution for problem one' do
    assert Euler.ProblemOne.sum_of_threes_and_fives(999) == 233168
  end
end
